core/modules/node/node.pages.inc

```diff
@@ -458,6 +513,208 @@ function node_form_submit($form, &$form_state) {
+    }
+
+    // In case this is the 'full' view mode, rename default.
+    if ($key == 'full') {
+      $view_mode_options[$key] = t('Default');
+    }
+  }
+
+  $path = 'node/add/' . $node->bundle();
+  if (!empty($node->nid)) {
+    $path = 'node/' . $node->nid . '/edit';
+  }
+  $query = array('node_tempstore_id' => $node_tempstore_id);
+  $url = url($path, array('query' => $query, 'absolute' => TRUE));
+
+  // Get the view mode to render the preview in.
+  $view_mode = empty($_GET['view_mode']) ? 'full' : $_GET['view_mode'];
+
+  $form['node'] = array(
+    '#type' => 'value',
+    '#value' => $node,
+  );
+
+  $form['node_tempstore_id'] = array(
+    '#type' => 'value',
+    '#value' => $node_tempstore_id,
+  );
+
+  $form['backlink'] = array(
+    '#type' => 'link',
+    '#title' => t('Back to content editing'),
+    '#href' => $url,
+    '#attributes' => array('class' => array('node-preview-backlink')),
+  );
+
+  $form['view_mode'] = array(
+    '#type' => 'select',
+    '#title' => t('Select a display mode'),
+    //'#title_display' => 'invisible',
+    '#options' => $view_mode_options,
+    '#default_value' => $view_mode,
+    '#attributes' => array(
+      'onChange' => 'this.form.submit()',
+      'class' => array('node-preview-display-mode', 'preview-button'),
+    )
+  );
+  // This must be the first submit function so the 'onChange' event in 
+  // calls it before any other submit.$form['view_mode']
+  $form['view_mode_submit'] = array(
+    '#type' => 'submit',
+    '#name' => 'op',
+    '#submit' => array('node_preview_banner_form_view_mode_submit'),
+    '#value' => t('Switch'),
+    '#attributes' => array(
+      'class' => array('node-preview-switch-button', 'preview-button', 'js-hide')
+    )
+  );
+  $form['preview_node_submit'] = array(
+    '#type' => 'submit',
+    '#submit' => array('node_preview_banner_form_node_submit'),
+    '#value' => t('Save'),
+    '#attributes' => array(
+      'class' => array('node-preview-save-button', 'preview-button'),
+    )
+  );
+  $form['#theme'] = 'node_preview_banner_form';
+
+  return $form;
+}
+
+/**
+ * Submit handler for the node preview view mode selection form.
+ *
+ * @see node_preview_banner_form()
+ */
+function node_preview_banner_form_node_submit(array $form, array &$form_state) {
+  form_load_include($form_state, 'inc', 'node', 'node.pages');
+  $node = $form_state['values']['node'];
+  node_object_prepare($node);
+  $type = $node->type;
+  $build = array();
+  $build['node'] = $node;
+  $build['values']['status'] = $node->status;
+  $build['values']['scheduled'] = $node->scheduled;
+  $build['values']['op'] = t('Save');
+  backdrop_form_submit($type . '_node_form', $build, $node);
+  $errors = form_get_errors();
+  if (!empty($errors)) {
+    foreach ($errors as $field_name => $message) {
+      watchdog('node', '%field: %message', array('%message'=> $message, '%field' => $field_name));
+    }
+  }
+  else {
+    backdrop_goto('node/' . $node->nid);
```

Comment:
I haven't tested your function for this, but `node_form_submit()` has more logic about determining if the user has access to view content first before sending them to the node. It seems like someone who doesn't have access to view will see an "access denied" once they save from the preview page.

Reply:
Fixed